### DISCUSSION TITLE
git plugin + git completions no workey

### DISCUSSION THREAD

**Discussion by jalelegenda** [ohmybash/oh-my-bash#633] [General]
If I enable git plugin and I enable git completions, completions do not work with git aliases. E.g. If I type in gco and hit tab I will not get a list of branches to checkout. Is this expected?

**Comment by akinomyoga** [comment]
> If I enable git plugin and I enable git completions, completions do not work with git aliases. E.g. If I type in gco and hit tab I will not get a list of branches to checkout. Is this expected?

I assume you talk about the shell aliases for the git command (but not the git aliases set up by e.g. `git config alias.gco`). It's expected that you do not get the completion unless you set up the git completion. Did you set up the git completion for your aliases? If not, please check [the instruction in a code comment in `git-completion.bash`](https://github.com/git/git/blob/25b0f41288718625b18495de23cc066394c09a92/contrib/completion/git-completion.bash#L41-L55).

**Comment by afintos-bo** [comment]
I can add a little more information about what I did to get it to work. Hopefully, that will help you. As akinomyoga suggested you have to add the completion yourself. I had to try this in a couple of places before I found something that worked that I was comfortable with. 

1. I am assuming you are using the standard custom config directory, if not make sure you have updated the OSH_CUSTOM environment variable. 
2. It looks like oh-my-bash sources files under the $OSH_CUSTOM directory, but there are also subdirectories (i.e. completions, etc) it does not appear to source by default. I created $OSH_CUSTOM/completions/git-completion.sh with the following line,

    ```
    __git_complete gco _git_checkout
   ```

You could also just add,

    
    __git_complete gco __git_main
    

or, if you are using "g" as a global leader for all of your git aliases you could add,

    __git_complete g __git_main

to get all aliases working at once (up to you). Either way you have to define what you want completion for in the git-completion.sh file.

3. After I added $OSH_CUSTOM/completions/git-completion.sh git completion still was not working for me. I noticed that oh-my-bash sources the top level *.{sh,bash} files in the $OSH_CUSTOM directory, but for some reason it was not sourcing the customization files in the subdirectories (maybe by design?). I added a new file in the $OSH_CUSTOM directory called customizations.sh that just has the following line,
    
    ```
    source "$OSH"/custom/completions/git-completion.sh
    ````

Completion is working with aliases I have defined now.

**Comment by codernirdesh** [comment]
Merging #659 fixes this.
